### PR TITLE
ZeroFox Integration

### PR DESCRIPTION
**Pull request by guyfreund**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/16314

## Required version of Demisto
any

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [ ] Code Review

### CODE REVIEW

File: Integrations/ZeroFox/ZeroFox.py [hunk 1]
```diff
@@ -0,0 +1,510 @@
+from distutils.util import strtobool
+from typing import Dict
+
+import demistomock as demisto
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: add `from CommonServerUserPython import *`
make sure these are the first three lines

File: Integrations/ZeroFox/ZeroFox.py [hunk 2]
```diff
@@ -0,0 +1,510 @@
+from distutils.util import strtobool
+from typing import Dict
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: move this to be under IMPORTS

File: Integrations/ZeroFox/ZeroFox.py [hunk 3]
```diff
@@ -0,0 +1,510 @@
+from distutils.util import strtobool
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: remove

File: Integrations/ZeroFox/ZeroFox.py [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+
+# Disable insecure warnings
+requests.packages.urllib3.disable_warnings()
+
+''' GLOBALS/PARAMS '''
+
+USERNAME = None
+PASSWORD = None
+TOKEN = None
+# # Remove trailing slash to prevent wrong URL path to service
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: remove comments

File: Integrations/ZeroFox/ZeroFox.py [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+import requests
+
+# Disable insecure warnings
+requests.packages.urllib3.disable_warnings()
+
+''' GLOBALS/PARAMS '''
+
+USERNAME = None
+PASSWORD = None
+TOKEN = None
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: shouldnt be global

File: Integrations/ZeroFox/ZeroFox.py [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+        'Alert Date': contents.get('Timestamp', ''),
+        'Status': contents.get('Status', '').title(),
+        'Source': contents.get('Network', '').title(),
+        'Rule': contents.get('RuleName'),
+        'Severity': contents.get('Severity'),
+        'Notes': contents.get('Notes') if contents.get('Notes') != '' else None,
+        'Tags': contents.get('Tags')
+    }
+
+def get_authorization_token():
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: this should return the token
use integration context to store the token

File: Integrations/ZeroFox/ZeroFox.py [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+    global TOKEN, HEADERS
+    TOKEN = request_response.get('token')
+    HEADERS = {
+        'Authorization': f'Token {TOKEN}',
+        'Content-Type': 'application/json',
+        'Accept': 'application/json'
+    }
+
+
+def http_request(method: str, url_suffix: str, params=None, data=None):
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: headers should be init here
call `get_authorization_token` from here

File: Integrations/ZeroFox/ZeroFox.py [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+        if demisto.command() == 'fetch-incidents':
+            LOG(error_msg)
+            LOG.print_log()
+            raise
+        else:
+            return_error(error_msg)
+
+
+# python2 uses __builtin__ python3 uses builtins
+if __name__ == '__builtin__' or __name__ == 'builtins':
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]:
```suggestion
if __name__ == 'builtins':
```

File: Integrations/ZeroFox/ZeroFox.py [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+        error_msg = str(e)
+        if demisto.command() == 'fetch-incidents':
+            LOG(error_msg)
+            LOG.print_log()
+            raise
+        else:
+            return_error(error_msg)
+
+
+# python2 uses __builtin__ python3 uses builtins
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: remove comment

File: Integrations/ZeroFox/ZeroFox.py [hunk 10]
```diff
@@ -0,0 +1,510 @@
...
+
+''' COMMANDS MANAGER / SWITCH PANEL '''
+
+''' EXECUTION '''
+
+
+def main():
+    LOG('Command being called is %s' % (demisto.command()))
+    try:
+        if USERNAME is None or PASSWORD is None or BASE_URL is None or USE_SSL is None:
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: remove condition

File: Integrations/ZeroFox/ZeroFox.py [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+            initialize_preset()
+        if TOKEN is None:
+            get_authorization_token()
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: remove from here

File: Integrations/ZeroFox/ZeroFox.yml [hunk 12]
```diff
@@ -0,0 +1,680 @@
...
+    - contextPath: ZeroFox.Alert.Asset.Image
+      description: The Asset Image.
+      type: String
+    - contextPath: ZeroFox.Alert.Status
+      description: Status of alert. Possible values are Open, Closed, Takedown:Accepted,
+        Takedown:Denied, Takedown:Requested and Whitelisted.
+      type: String
+    - contextPath: ZeroFox.Alert.Timestamp
+      description: ISO-8601 date-time string of when alert was created.
+      type: String
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]: should be type date

File: Integrations/ZeroFox/ZeroFox.yml [hunk 13]
```diff
@@ -0,0 +1,680 @@
...
+    - contextPath: ZeroFox.Alert.RuleName
+      description: Name of rule for which alert was created on. null if rule has been
+        deleted.
+      type: String
+    - contextPath: ZeroFox.Alert.LastModified
+      description: The datetime in which the alert was last modified.
+      type: String
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]: should be type date

File: Integrations/ZeroFox/ZeroFox.yml [hunk 14]
```diff
@@ -0,0 +1,680 @@
...
+      isArray: false
+      name: alert_id
+      required: false
+      secret: false
+    - auto: PREDEFINED
+      default: false
+      description: Severity level of alert. 1 - 5 (Critical).
+      isArray: false
+      name: severity
+      predefined:
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]: change to human readable severties

File: Integrations/ZeroFox/ZeroFox.yml [hunk 15]
```diff
@@ -0,0 +1,680 @@
...
+      description: CSV of alert IDs
+      isArray: false
+      name: alert_id
+      required: false
+      secret: false
+    - auto: PREDEFINED
+      default: false
+      description: Severity level of alert. 1 - 5 (Critical).
+      isArray: false
+      name: severity
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]: change to ui name (risk score)

File: Integrations/ZeroFox/ZeroFox.py [hunk 16]
```diff
@@ -0,0 +1,510 @@
...
+    )
+    # Handle error responses gracefully
+    if res.status_code not in {200, 201}:
+        err_msg: str = f'Error in ZeroFox Integration API call [{res.status_code}] - {res.reason}\n'
+        try:
+            res_json = res.json()
+            if 'error' in res_json:
+                err_msg += res_json.get('error')
+            else:
+                err_msg += res_json
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: concat str and dict, verify

File: Integrations/ZeroFox/ZeroFox.py [hunk 17]
```diff
@@ -0,0 +1,510 @@
...
+        data=data,
+        headers=HEADERS
+    )
+    # Handle error responses gracefully
+    if res.status_code not in {200, 201}:
+        err_msg: str = f'Error in ZeroFox Integration API call [{res.status_code}] - {res.reason}\n'
+        try:
+            res_json = res.json()
+            if 'error' in res_json:
+                err_msg += res_json.get('error')
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: verify this is a str

File: Integrations/ZeroFox/ZeroFox.py [hunk 18]
```diff
@@ -0,0 +1,510 @@
...
+            else:
+                err_msg += res_json
+        except ValueError:
+            pass
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: need to concat here `res.text` or `res.content`

File: Integrations/ZeroFox/ZeroFox.py [hunk 19]
```diff
@@ -0,0 +1,510 @@
...
+        finally:
+            return_error(err_msg)
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]:
```suggestion
            raise ValueError(err_msg
```

File: Integrations/ZeroFox/ZeroFox.yml [hunk 20]
```diff
@@ -0,0 +1,680 @@
...
+      description: The ID of the alert to be assigned.
+      isArray: false
+      name: alert_id
+      required: true
+      secret: false
+    - default: false
+      description: The email of the user to assign the alert to.
+      isArray: false
+      name: subject_email
+      required: true
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]: not required

File: Integrations/ZeroFox/ZeroFox.yml [hunk 21]
```diff
@@ -0,0 +1,680 @@
...
+      type: Unknown
+  - arguments:
+    - default: false
+      description: The ID of the alert to be assigned.
+      isArray: false
+      name: alert_id
+      required: true
+      secret: false
+    - default: false
+      description: The email of the user to assign the alert to.
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]: improve description

File: Integrations/ZeroFox/ZeroFox.yml [hunk 22]
```diff
@@ -0,0 +1,680 @@
...
+      isArray: false
+      name: subject_email
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]:
```suggestion
      name: email
```

File: Integrations/ZeroFox/ZeroFox.yml [hunk 23]
```diff
@@ -0,0 +1,680 @@
...
+      required: true
+      secret: false
+    - default: false
+      description: The name of the user being assigned.
+      isArray: false
+      name: subject_name
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]:
```suggestion
      name: username
```

File: Integrations/ZeroFox/ZeroFox.py [hunk 24]
```diff
@@ -0,0 +1,510 @@
...
+    response_content = http_request('POST', url_suffix, data=json.dumps(request_body))
+    return response_content
+
+
+def alert_user_assignment_command():
+    alert_id: int = demisto.args().get('alert_id')
+    subject_name: str = demisto.args().get('subject_name')
+    subject_email: str = demisto.args().get('subject_email')
+    alert_user_assignment(alert_id, subject_email, subject_name)
+    context = get_alert_context_no_war_room(alert_id)
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: push to context only id and assignee

File: Integrations/ZeroFox/ZeroFox.py [hunk 25]
```diff
@@ -0,0 +1,510 @@
...
+    return_outputs(
+        f'User: {subject_email} has been assigned to Alert: {alert_id} successfully.',
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]:
```suggestion
        f'User {subject_email} has been assigned to alert {alert_id} successfully.',
```

File: Integrations/ZeroFox/ZeroFox.yml [hunk 26]
```diff
@@ -0,0 +1,680 @@
...
+      required: true
+      secret: false
+    deprecated: false
+    description: Assigns an alert to a user
+    execution: false
+    name: zerofox-alert-user-assignment
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]: add outputs id and assignee

File: Integrations/ZeroFox/ZeroFox.yml [hunk 27]
```diff
@@ -0,0 +1,680 @@
...
+      defaultValue: add
+      isArray: false
+      name: action
+      predefined:
+      - add
+      - remove
+      required: false
+      secret: false
+    - default: false
+      description: Unique ID of the alert.
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]: where to get?

File: Integrations/ZeroFox/ZeroFox.py [hunk 28]
```diff
@@ -0,0 +1,510 @@
...
+                'alert': alert_id
+            }
+        ]
+    }
+    response_content = http_request('POST', url_suffix, data=json.dumps(request_body))
+    return response_content
+
+
+def modify_alert_tags_command():
+    alert_id = demisto.args().get('alert_id')
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: verify with alerts id which is not found

File: Integrations/ZeroFox/ZeroFox.py [hunk 29]
```diff
@@ -0,0 +1,510 @@
...
+    action_string = demisto.args().get('action')
+    action = True if action_string == 'add' else False
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: remove this condtion

File: Integrations/ZeroFox/ZeroFox.py [hunk 30]
```diff
@@ -0,0 +1,510 @@
...
+        f'User: {subject_email} has been assigned to Alert: {alert_id} successfully.',
+        context,
+        raw_response={}
+    )
+
+
+def modify_alert_tags(alert_id, action, tags_list_string):
+    url_suffix: str = '/alerttagchangeset/'
+    tags_list_name: str = 'added' if action else 'removed'
+    tags_list: list = tags_list_string.split(',')
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: use `argsToList`

File: Integrations/ZeroFox/ZeroFox.py [hunk 31]
```diff
@@ -0,0 +1,510 @@
...
+
+def modify_alert_tags_command():
+    alert_id = demisto.args().get('alert_id')
+    action_string = demisto.args().get('action')
+    action = True if action_string == 'add' else False
+    tags_list_string = demisto.args().get('tags')
+    modify_alert_tags(alert_id, action, tags_list_string)
+    context = get_alert_context_no_war_room(alert_id)
+    return_outputs(
+        'Changes were successfully made.',
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: improve message

File: Integrations/ZeroFox/ZeroFox.py [hunk 32]
```diff
@@ -0,0 +1,510 @@
...
+        context,
+        raw_response={}
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: return context here

File: Integrations/ZeroFox/ZeroFox.py [hunk 33]
```diff
@@ -0,0 +1,510 @@
...
+    incident = {
+        'rawJSON': json.dumps(alert),
+        'name': 'ZeroFox Alert ' + str(alert.get('id')),  # not sure if it's the right name
+        'occurred': alert.get('timestamp')  # not sure if it's the right field
+    }
+    return incident
+
+
+# return context convention of alert without printing to war room
+def get_alert_context_no_war_room(alert_id):
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: improve func name

File: Integrations/ZeroFox/ZeroFox.py [hunk 34]
```diff
@@ -0,0 +1,510 @@
...
+    alert = get_alert(alert_id).get('alert')
+    if not alert:
+        return {}
+    contents = get_alert_contents(alert)
+    context = {'ZeroFox.Alert(val.ID && val.ID === obj.ID)': contents}
+    return context
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: return `contents` and handle context path in the functions

File: Integrations/ZeroFox/ZeroFox.yml (hunk 15 above)
Comment (guyfreund) [reply Integrations/ZeroFox/ZeroFox.yml]: Changed to Risk Rating

File: Integrations/ZeroFox/ZeroFox.py (hunk 16 above)
Comment (guyfreund) [reply Integrations/ZeroFox/ZeroFox.py]: should I delete res_json?

File: Integrations/ZeroFox/ZeroFox.py (hunk 32 above)
Comment (guyfreund) [reply Integrations/ZeroFox/ZeroFox.py]: returned contents, I think it's the right thing to return here

File: Integrations/ZeroFox/ZeroFox.py (hunk 6 above)
Comment (guyfreund) [reply Integrations/ZeroFox/ZeroFox.py]: TODO: use integration context to store the token

File: Integrations/ZeroFox/ZeroFox.yml (hunk 21 above)
Comment (guyfreund) [reply Integrations/ZeroFox/ZeroFox.yml]: deleted the email arg

File: Integrations/ZeroFox/ZeroFox.yml (hunk 22 above)
Comment (guyfreund) [reply Integrations/ZeroFox/ZeroFox.yml]: deleted the email arg

File: Integrations/ZeroFox/ZeroFox.py [hunk 35]
```diff
@@ -0,0 +1,579 @@
...
+        return_outputs('Unexpected outputs from API call.', outputs={})
+
+
+def list_entities(params: Dict):
+    url_suffix: str = '/entities/'
+    response_content: Dict = http_request('GET', url_suffix, params=params)
+    return response_content
+
+
+def list_entities_command():
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: should be able to filter only by whats available in the ui and all filters should be by human readable and not IDs

File: Integrations/ZeroFox/ZeroFox.py [hunk 36]
```diff
@@ -0,0 +1,579 @@
...
+    response_content: Dict = list_entities(params)
+    if not response_content:
+        return_outputs('No entities found.', outputs={})
+    elif isinstance(response_content, Dict):
+        entities: List = response_content.get('entities')
+        contents: Dict = [get_entity_contents(entity) for entity in entities]
+        contents_war_room: Dict = [get_entity_contents_war_room(content) for content in contents]
+        context: Dict = {'ZeroFox.Entity(val.ID && val.ID === obj.ID)': contents}
+        return_outputs(
+            tableToMarkdown('ZeroFox Entities', contents_war_room, removeNull=True),
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: if no results, dont return a table

File: Integrations/ZeroFox/ZeroFox.yml [hunk 37]
```diff
@@ -0,0 +1,938 @@
...
+    - default: false
+      description: indicating type of string matching for comparing name to impersonators
+      isArray: false
+      name: strict_name_matching
+      required: false
+      secret: false
+    - default: false
+      description: profile image for entity
+      isArray: false
+      name: image
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]: remove

File: Integrations/ZeroFox/ZeroFox.yml [hunk 38]
```diff
@@ -0,0 +1,938 @@
...
+    - default: false
+      description: Name of the entity (may be non-unique)
+      isArray: false
+      name: name
+      required: true
+      secret: false
+    - default: false
+      description: indicating type of string matching for comparing name to impersonators
+      isArray: false
+      name: strict_name_matching
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]: add predefined values

File: Integrations/ZeroFox/ZeroFox.yml [hunk 39]
```diff
@@ -0,0 +1,938 @@
...
+      name: image
+      required: false
+      secret: false
+    - default: false
+      description: |-
+        List of string labels for tagging the entity.
+        Seperated by a comma, for example:
+        label1,label2,label3
+      isArray: false
+      name: labels
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]: change to tags

File: Integrations/ZeroFox/ZeroFox.py [hunk 40]
```diff
@@ -0,0 +1,579 @@
...
+    labels: List = argToList(labels, ',')
+    policy: int = int(demisto.args().get('policy'))
+    organization: str = demisto.args().get('organization')
+    response_content: Dict = create_entity(name, strict_name_matching, image, labels, policy, organization)
+    if not response_content or not isinstance(response_content, Dict):
+        raise Exception('Unexpected outputs from API call.')
+    entity_id: int = response_content.get('id')
+    return_outputs(
+        f'Entity has been created successfully. ID: {entity_id}',
+        {'ZeroFox.Entity(val.ID && val.ID === obj.ID)': {'ID': entity_id}},
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: check context returned and build context

File: Integrations/ZeroFox/ZeroFox.py [hunk 41]
```diff
@@ -0,0 +1,579 @@
...
+    labels: str = demisto.args().get('args')
+    labels: List = argToList(labels, ',')
+    policy: int = int(demisto.args().get('policy'))
+    organization: str = demisto.args().get('organization')
+    response_content: Dict = create_entity(name, strict_name_matching, image, labels, policy, organization)
+    if not response_content or not isinstance(response_content, Dict):
+        raise Exception('Unexpected outputs from API call.')
+    entity_id: int = response_content.get('id')
+    return_outputs(
+        f'Entity has been created successfully. ID: {entity_id}',
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: improve message

File: Integrations/ZeroFox/ZeroFox.py [hunk 42]
```diff
@@ -0,0 +1,579 @@
...
+
+# TODO: REMEMBER TO DELETE
+def fetch_incidents_command():
+    return_outputs('fetch 1', outputs={})
+    fetch_incidents()
+    return_outputs('fetch 2', outputs={})
+    fetch_incidents()
+
+
+def fetch_incidents():
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: add typing

File: Integrations/ZeroFox/ZeroFox.py [hunk 43]
```diff
@@ -0,0 +1,579 @@
...
+    date_format = '%Y-%m-%dT%H:%M:%S'
+    last_run = demisto.getLastRun()
+    if last_run and last_run.get('last_fetched_event_timestamp'):
+        last_update_time = last_run['last_fetched_event_timestamp']
+    else:
+        last_update_time = parse_date_range(FETCH_TIME, date_format=date_format)[0]
+    incidents = []
+    limit = demisto.params().get('fetch_limit')
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]:
```suggestion
    limit = demisto.params().get('fetch_limit', 10)
```

File: Integrations/ZeroFox/ZeroFox.py [hunk 44]
```diff
@@ -0,0 +1,579 @@
...
+    response_content = list_alerts({'sort_direction': 'asc', 'limit': limit, 'min_timestamp': last_update_time})
+    alerts = response_content.get('alerts')
+    if not alerts:
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: remove

File: Integrations/ZeroFox/ZeroFox.py [hunk 45]
```diff
@@ -0,0 +1,579 @@
...
+    date_format = '%Y-%m-%dT%H:%M:%S'
+    last_run = demisto.getLastRun()
+    if last_run and last_run.get('last_fetched_event_timestamp'):
+        last_update_time = last_run['last_fetched_event_timestamp']
+    else:
+        last_update_time = parse_date_range(FETCH_TIME, date_format=date_format)[0]
+    incidents = []
+    limit = demisto.params().get('fetch_limit')
+    response_content = list_alerts({'sort_direction': 'asc', 'limit': limit, 'min_timestamp': last_update_time})
+    alerts = response_content.get('alerts')
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]:
```suggestion
    alerts = response_content.get('alerts', [])
```

File: Integrations/ZeroFox/ZeroFox.py [hunk 46]
```diff
@@ -0,0 +1,579 @@
...
+        return 'High'
+    elif severity_num == 5:
+        return 'Critical'
+
+
+# transforms an alert to incident convention
+def alert_to_incident(alert: Dict):
+    incident: Dict = {
+        'rawJSON': json.dumps(alert),
+        'name': 'ZeroFox Alert ' + str(alert.get('id')),
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: use format

File: Integrations/ZeroFox/ZeroFox.py [hunk 47]
```diff
@@ -0,0 +1,579 @@
...
+        return
+    for alert in alerts:
+        alert_id = alert.get('id')
+        ts = alert.get('timestamp')
+        # TODO: REMEMBER TO DELETE
+        return_outputs(f'Alert: {alert_id}, TS: {ts}', outputs={})
+        incident = alert_to_incident(alert)
+        incidents.append(incident)
+    # max_update_time is the timestamp of the last alert in alerts (alerts is a sorted list)
+    last_alert_timestamp = str(alerts[len(alerts) - 1].get('timestamp'))
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]:
```suggestion
    last_alert_timestamp = str(alerts[-1].get('timestamp'))
```

File: Integrations/ZeroFox/ZeroFox.py [hunk 48]
```diff
@@ -0,0 +1,579 @@
...
+    if '+' in last_alert_timestamp:
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: add comment here

File: Integrations/ZeroFox/ZeroFox.py [hunk 49]
```diff
@@ -0,0 +1,579 @@
...
+        max_update_time = last_alert_timestamp.split('+')[0]
+    else:
+        max_update_time = last_alert_timestamp.split('-')[0]
+    # add 1 second to last alert timestamp, in order to prevent duplicated alerts
+    max_update_time = (datetime.strptime(max_update_time, date_format) + timedelta(0, 1)).isoformat()
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]:
```suggestion
    max_update_time = (datetime.strptime(max_update_time, date_format) + timedelta(seconds=1)).isoformat()
```

File: Integrations/ZeroFox/ZeroFox.py [hunk 50]
```diff
@@ -0,0 +1,579 @@
...
+    response_content: Dict = list_alerts(params)
+    if not response_content:
+        return_outputs('No alerts found.', outputs={})
+    elif isinstance(response_content, Dict):
+        alerts: List = response_content.get('alerts')
+        contents: Dict = [get_alert_contents(alert) for alert in alerts]
+        contents_war_room: Dict = [get_alert_contents_war_room(content) for content in contents]
+        context: Dict = {'ZeroFox.Alert(val.ID && val.ID === obj.ID)': contents}
+        return_outputs(
+            tableToMarkdown('ZeroFox Alerts', contents_war_room, removeNull=True),
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: dont return empty table if no results

**Comment by guyfreund** [comment]
@kirbles19 Hey Alex, can you please review the [ZeroFox.yml](https://github.com/demisto/content/edit/zerofox-integration/Integrations/ZeroFox/ZeroFox.yml?pr=%2Fdemisto%2Fcontent%2Fpull%2F4111) file?

File: Integrations/ZeroFox/ZeroFox.py [hunk 51]
```diff
@@ -0,0 +1,579 @@
...
+    url_suffix: str = '/api-token-auth/'
+    data_for_request: Dict = {
+        'username': USERNAME,
+        'password': PASSWORD
+    }
+    response_content: Dict = http_request('POST', url_suffix, data=data_for_request, continue_err=True,
+                                          regular_request=False)
+    if not response_content or not isinstance(response_content, Dict):
+        raise Exception('Unexpected outputs from API call.')
+    token: str = response_content.get('token')
```
[the remark this answers is not in the document]
Comment (guyfreund) [reply Integrations/ZeroFox/ZeroFox.py]: Do I need to do also:
> name: str = demisto.args().get('name', '')

or just:
> name: str = demisto.args().get('name')

File: Integrations/ZeroFox/ZeroFox.py [hunk 52]
```diff
@@ -0,0 +1,579 @@
...
+                err_msg += res.content
+            finally:
+                raise ValueError(err_msg)
+        else:
+            try:
+                return res.json()
+            except ValueError:
+                return res.content
+    except requests.exceptions.ConnectTimeout:
+        err_msg: str = 'Connection Timeout Error - potential reasons may be that the Server URL parameter' \
```
[the remark this answers is not in the document]
Comment (guyfreund) [reply Integrations/ZeroFox/ZeroFox.py]: image

File: Integrations/ZeroFox/ZeroFox.py (hunk 50 above)
Comment (guyfreund) [reply Integrations/ZeroFox/ZeroFox.py]: What did you mean here?

**Comment by guyfreund** [comment]
Hey @kirbles19 these are the error messages presented to the user in the python file. Please review the following lines:
242
285
297
302-303
306-307
310-311
317-319
342
364
386
408
433
471
536
624
627
664

**Comment by kirbles19** [comment]
@ygerowitz Review this file: https://github.com/demisto/content/blob/69de3dce674b5fd1b784da5b69f66e7c27a2170e/Integrations/ZeroFox/ZeroFox.yml

File: Integrations/ZeroFox/ZeroFox.yml [hunk 53]
```diff
@@ -0,0 +1,970 @@
...
+  display: First fetch timestamp (<number> <time unit>, e.g., 12 hours, 7 days)
+  name: fetch_time
+  required: false
+  type: 0
+- defaultvalue: '10'
+  display: Fetch Limit
+  name: fetch_limit
+  required: true
+  type: 0
+description: ZeroFox Integration - Cloud-based SaaS to detect risks found on social
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.yml]:
```suggestion
description: Cloud-based SaaS to detect risks found on social
```

File: Integrations/ZeroFox/ZeroFox.py [hunk 54]
```diff
@@ -0,0 +1,758 @@
...
+        if params:
+            value: str = params.get(key, '')
+            if value:
+                params[key] = int(value)
+                return params[key]
+    except ValueError:
+        raise Exception(f'This value for {key} must be an integer.')
+
+
+# transforms severity number to string representation
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: please move all the comments you added above every function to the docstring

File: Integrations/ZeroFox/ZeroFox.py [hunk 55]
```diff
@@ -0,0 +1,758 @@
...
+        max_update_time = last_update_time
+        demisto.setLastRun({'last_fetched_event_timestamp': max_update_time})
+    demisto.incidents(incidents)
+
+
+def test_module():
+    """
+    Performs basic get request to get item samples
+    """
+    get_authorization_token()
```
Comment (Itay4) [inline Integrations/ZeroFox/ZeroFox.py]: this is already called from `http_request` which is called from `get_policy_types` so you can remove it from here

**Comment by ygerowitz** [comment]
Reviewed ZeroFOX.yml file.

**Pull request merged**